File: sidecar/translator_sidecar/provider_contract.py
"""Versioned provider contracts shared with the Rust daemon."""

from __future__ import annotations

from enum import Enum
from typing import Literal, TypedDict
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator


class ContractModel(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)


class AudioDirection(str, Enum):
    MICROPHONE = "microphone"
    SPEAKER = "speaker"


class TranslationMode(str, Enum):
    QUALITY_FIRST = "quality_first"
    BALANCED = "balanced"
    STREAMING_FIRST = "streaming_first"


class Language(str, Enum):
    RU = "ru"
    EN = "en"


class VoiceGender(str, Enum):
    MALE = "male"
    FEMALE = "female"


class VoiceEngine(str, Enum):
    PIPER = "piper"
    SILERO = "silero"
    OPENAI = "openai"


class SampleFormat(str, Enum):
    S16LE = "s16le"


class ProviderId(str, Enum):
    LOCAL = "local"
    OPENAI = "openai"


class SafeErrorCode(str, Enum):
    PROVIDER_UNAVAILABLE = "provider_unavailable"
    MODEL_NOT_LOADED = "model_not_loaded"
    UNSUPPORTED_LANGUAGE_PAIR = "unsupported_language_pair"
    QUEUE_OVERFLOW = "queue_overflow"
    CANCELLED = "cancelled"
    NO_SPEECH = "no_speech"
    CLOUD_NOT_ENABLED = "cloud_not_enabled"
    PROVIDER_AUTH_FAILED = "provider_auth_failed"


SAFE_ERROR_MESSAGES: dict[SafeErrorCode, str] = {
    SafeErrorCode.PROVIDER_UNAVAILABLE: "Provider is unavailable",
    SafeErrorCode.MODEL_NOT_LOADED: "Required model is not loaded",
    SafeErrorCode.UNSUPPORTED_LANGUAGE_PAIR: "Language pair is not supported",
    SafeErrorCode.QUEUE_OVERFLOW: "Provider queue limit was reached",
    SafeErrorCode.CANCELLED: "Provider operation was cancelled",
    SafeErrorCode.NO_SPEECH: "No speech was detected",
    SafeErrorCode.CLOUD_NOT_ENABLED: "Cloud provider is not enabled",
    SafeErrorCode.PROVIDER_AUTH_FAILED: "Provider authentication failed",
}


class VoiceProfile(ContractModel):
    language: Language
    gender: VoiceGender
    engine: VoiceEngine
    model_path: str | None = None
    provider_voice_id: str | None = None


class PcmFormat(ContractModel):
    sample_rate_hz: Literal[16000, 24000, 48000]
    channels: Literal[1, 2]
    sample_format: Literal[SampleFormat.S16LE]
    frame_duration_ms: Literal[20, 40, 60, 80, 100]


class OpenProviderSession(ContractModel):
    schema_version: Literal["translator.provider.open_session.v1"] = (
        "translator.provider.open_session.v1"
    )
    session_id: UUID
    provider_id: ProviderId
    direction_id: AudioDirection
    source_language: Language
    target_language: Language
    mode: TranslationMode
    requested_input_format: PcmFormat
    requested_output_format: PcmFormat
    voice_profile: VoiceProfile
    debug_text_enabled: bool = False


class ProviderCapabilities(ContractModel):
    audio_output: Literal[True] = True
    transcript_delta: bool
    translation_delta: bool
    cancellation: bool
    cloud_egress: bool


class ProviderSessionOpened(ContractModel):
    schema_version: Literal["translator.provider.session_opened.v1"] = (
        "translator.provider.session_opened.v1"
    )
    session_id: UUID
    direction_id: AudioDirection
    event_sequence: int = Field(ge=0)
    negotiated_input_format: PcmFormat
    negotiated_output_format: PcmFormat
    capabilities: ProviderCapabilities


class ProviderProbeRequest(ContractModel):
    schema_version: Literal["translator.provider.probe_request.v1"] = (
        "translator.provider.probe_request.v1"
    )


class ProviderProbeResponse(ContractModel):
    schema_version: Literal["translator.provider.probe_response.v1"] = (
        "translator.provider.probe_response.v1"
    )
    generation_id: UUID


class CloseRequestReason(str, Enum):
    USER_STOP = "user_stop"
    ROUTE_REMOVED = "route_removed"
    DEVICE_UNAVAILABLE = "device_unavailable"
    PROVIDER_SWITCH = "provider_switch"
    DAEMON_SHUTDOWN = "daemon_shutdown"


class SessionCloseReason(str, Enum):
    USER_STOP = "user_stop"
    ROUTE_REMOVED = "route_removed"
    DEVICE_UNAVAILABLE = "device_unavailable"
    PROVIDER_SWITCH = "provider_switch"
    DAEMON_SHUTDOWN = "daemon_shutdown"
    PROVIDER_FAILURE = "provider_failure"
    CLOSE_TIMEOUT = "close_timeout"


class CloseProviderSession(ContractModel):
    schema_version: Literal["translator.provider.close_session.v1"] = (
        "translator.provider.close_session.v1"
    )
    session_id: UUID
    reason: CloseRequestReason


class CancelReason(str, Enum):
    LATENCY_POLICY = "latency_policy"
    ROUTE_REMOVED = "route_removed"
    USER_INTERRUPT = "user_interrupt"
    QUEUE_OVERFLOW = "queue_overflow"


class CancelUtterance(ContractModel):
    schema_version: Literal["translator.provider.cancel_utterance.v1"] = (
        "translator.provider.cancel_utterance.v1"
    )
    session_id: UUID
    direction_id: AudioDirection
    utterance_id: UUID
    reason: CancelReason


class UpdateDebugText(ContractModel):
    schema_version: Literal["translator.provider.update_debug_text.v1"] = (
        "translator.provider.update_debug_text.v1"
    )
    session_id: UUID
    enabled: bool


class ProviderInputFrame(ContractModel):
    schema_version: Literal["translator.provider.input.v1"] = (
        "translator.provider.input.v1"
    )
    session_id: UUID
    direction_id: AudioDirection
    stream_id: UUID
    utterance_id: UUID
    sequence: int = Field(ge=0)
    capture_monotonic_ns: int = Field(ge=0)
    sample_rate_hz: Literal[16000, 24000, 48000]
    channels: Literal[1, 2]
    sample_format: Literal[SampleFormat.S16LE]
    frame_duration_ms: Literal[20, 40, 60, 80, 100]
    source_language: Language
    target_language: Language
    mode: TranslationMode
    pcm: bytes
    end_of_utterance: bool


class ProviderAudioDelta(ContractModel):
    schema_version: Literal["translator.provider.audio_delta.v1"] = (
        "translator.provider.audio_delta.v1"
    )
    session_id: UUID
    direction_id: AudioDirection
    stream_id: UUID
    utterance_id: UUID
    sequence: int = Field(ge=0)
    event_sequence: int = Field(ge=0)
    provider_monotonic_ns: int = Field(ge=0)
    sample_rate_hz: Literal[16000, 24000, 48000]
    channels: Literal[1, 2]
    sample_format: Literal[SampleFormat.S16LE]
    frame_duration_ms: Literal[20, 40, 60, 80, 100]
    pcm: bytes


class ProviderTranscriptDelta(ContractModel):
    schema_version: Literal["translator.provider.transcript_delta.v1"] = (
        "translator.provider.transcript_delta.v1"
    )
    session_id: UUID
    direction_id: AudioDirection
    stream_id: UUID
    utterance_id: UUID
    event_sequence: int = Field(ge=0)
    text: str
    is_final: bool


class ProviderTranslationDelta(ContractModel):
    schema_version: Literal["translator.provider.translation_delta.v1"] = (
        "translator.provider.translation_delta.v1"
    )
    session_id: UUID
    direction_id: AudioDirection
    stream_id: UUID
    utterance_id: UUID
    event_sequence: int = Field(ge=0)
    text: str
    stable_prefix: bool
    is_final: bool


class UtteranceOutcome(str, Enum):
    COMPLETED = "completed"
    CANCELLED = "cancelled"
    DROPPED = "dropped"


class ProviderUtteranceFinal(ContractModel):
    schema_version: Literal["translator.provider.utterance_final.v1"] = (
        "translator.provider.utterance_final.v1"
    )
    session_id: UUID
    direction_id: AudioDirection
    stream_id: UUID
    utterance_id: UUID
    event_sequence: int = Field(ge=0)
    final_audio_sequence: int | None = Field(default=None, ge=0)
    outcome: UtteranceOutcome


class ProviderSessionClosed(ContractModel):
    schema_version: Literal["translator.provider.session_closed.v1"] = (
        "translator.provider.session_closed.v1"
    )
    session_id: UUID
    direction_id: AudioDirection
    event_sequence: int = Field(ge=0)
    reason: SessionCloseReason


class ProviderState(str, Enum):
    STARTING = "starting"
    READY = "ready"
    DEGRADED = "degraded"
    BACKPRESSURE = "backpressure"
    RESTARTING = "restarting"
    UNAVAILABLE = "unavailable"
    CLOSED = "closed"


class ModelKind(str, Enum):
    ASR = "asr"
    MT = "mt"
    TTS = "tts"
    SPEECH_TO_SPEECH = "speech_to_speech"


class ModelState(str, Enum):
    NOT_LOADED = "not_loaded"
    LOADING = "loading"
    READY = "ready"
    FAILED = "failed"


class ComputeDevice(str, Enum):
    CUDA = "cuda"
    CPU = "cpu"
    CLOUD = "cloud"


class ModelHealth(ContractModel):
    kind: ModelKind
    id: str
    state: ModelState
    device: ComputeDevice | None = None
    safe_error_code: str | None = None


class ProviderQueues(ContractModel):
    provider_input_buffered_ms: int = Field(ge=0)
    provider_output_buffered_ms: int = Field(ge=0)
    queue_lag_ms: int = Field(ge=0)


class ProviderRetry(ContractModel):
    attempt: int = Field(ge=0)
    next_retry_after_ms: int = Field(ge=0)
    reason_code: str


class SafeErrorSummary(ContractModel):
    code: SafeErrorCode
    message: str
    retryable: bool

    @model_validator(mode="after")
    def validate_static_message(self) -> SafeErrorSummary:
        if self.message != SAFE_ERROR_MESSAGES[self.code]:
            raise ValueError("message must be the static message for code")
        return self


class ProviderHealth(ContractModel):
    schema_version: Literal["translator.provider.health.v1"] = (
        "translator.provider.health.v1"
    )
    session_id: UUID
    direction_id: AudioDirection
    event_sequence: int = Field(ge=0)
    provider_id: ProviderId
    provider_name: str
    state: ProviderState
    models: tuple[ModelHealth, ...]
    queues: ProviderQueues
    retry: ProviderRetry | None = None
    safe_error: SafeErrorSummary | None = None


class ProviderLatency(ContractModel):
    schema_version: Literal["translator.provider.latency.v1"] = (
        "translator.provider.latency.v1"
    )
    session_id: UUID
    direction_id: AudioDirection
    stream_id: UUID
    event_sequence: int = Field(ge=0)
    utterance_id: UUID | None = None
    asr_first_text_ms: int | None = Field(default=None, ge=0)
    asr_final_text_ms: int | None = Field(default=None, ge=0)
    mt_first_text_ms: int | None = Field(default=None, ge=0)
    tts_first_audio_ms: int | None = Field(default=None, ge=0)
    provider_total_ms: int | None = Field(default=None, ge=0)


class LatencyPolicyState(ContractModel):
    direction_id: AudioDirection
    current_mode: TranslationMode
    rolling_window_seconds: Literal[60] = 60
    minimum_samples: Literal[20] = 20
    degrade_after_consecutive_windows: Literal[2] = 2
    recover_after_consecutive_windows: Literal[5] = 5
    cooldown_seconds_after_change: Literal[120] = 120
    p95_first_audio_ms: int = Field(ge=0)
    p95_last_audio_ms: int = Field(ge=0)
    p95_queue_lag_ms: int = Field(ge=0)
    last_mode_change_at: str | None = None
    reason: str | None = None


class PrivacySafeProviderError(ContractModel):
    schema_version: Literal["translator.provider.error.v1"] = (
        "translator.provider.error.v1"
    )
    session_id: UUID
    direction_id: AudioDirection
    stream_id: UUID | None = None
    utterance_id: UUID | None = None
    event_sequence: int = Field(ge=0)
    code: SafeErrorCode
    retryable: bool
    safe_message: str = Field(max_length=160)

    @model_validator(mode="after")
    def validate_static_message(self) -> PrivacySafeProviderError:
        if self.safe_message != SAFE_ERROR_MESSAGES[self.code]:
            raise ValueError("safe_message must be the static message for code")
        return self


def make_provider_error(
    *,
    session_id: UUID,
    direction_id: AudioDirection,
    event_sequence: int,
    code: SafeErrorCode,
    retryable: bool,
    stream_id: UUID | None = None,
    utterance_id: UUID | None = None,
) -> PrivacySafeProviderError:
    return PrivacySafeProviderError(
        session_id=session_id,
        direction_id=direction_id,
        event_sequence=event_sequence,
        stream_id=stream_id,
        utterance_id=utterance_id,
        code=code,
        retryable=retryable,
        safe_message=SAFE_ERROR_MESSAGES[code],
    )


class ProviderLogFields(TypedDict):
    event: Literal["provider_error"]
    code: str
    retryable: bool


def provider_log_fields(error: PrivacySafeProviderError) -> ProviderLogFields:
    """Project an error to fields permitted in normal operational logs."""
    return {
        "event": "provider_error",
        "code": error.code.value,
        "retryable": error.retryable,
    }

File: crates/translator-core/tests/provider_contract.rs
use translator_core::{
    AudioDirection, CloseProviderSession, CloseRequestReason, CloseSessionVersion, Language,
    LatencyPolicyState, ModelHealth, ModelKind, ModelState, OpenProviderSession,
    OpenSessionVersion, PcmFormat, PrivacySafeLogEvent, PrivacySafeProviderError,
    ProviderAudioDelta, ProviderAudioDeltaVersion, ProviderCapabilities, ProviderHealth,
    ProviderHealthVersion, ProviderId, ProviderInputFrame, ProviderInputVersion, ProviderLatency,
    ProviderLatencyVersion, ProviderProbeRequest, ProviderProbeRequestVersion,
    ProviderProbeResponse, ProviderProbeResponseVersion, ProviderQueues, ProviderSessionClosed,
    ProviderSessionOpened, ProviderState, RequiredTrue, SafeErrorCode, SampleFormat,
    SessionCloseReason, SessionClosedVersion, SessionOpenedVersion, TranslationMode,
    UpdateDebugText, UpdateDebugTextVersion, VoiceEngine, VoiceGender, VoiceProfile,
};
use uuid::Uuid;

fn pcm_format() -> PcmFormat {
    PcmFormat::try_new(16_000, 1, SampleFormat::S16Le, 40).expect("valid PCM format")
}

#[test]
fn provider_health_matches_the_versioned_wire_shape() {
    let health = ProviderHealth {
        schema_version: ProviderHealthVersion::V1,
        session_id: Uuid::new_v4(),
        direction_id: AudioDirection::Microphone,
        event_sequence: 7,
        provider_id: ProviderId::Local,
        provider_name: "local-cascade".to_owned(),
        state: ProviderState::Ready,
        models: vec![ModelHealth {
            kind: ModelKind::Asr,
            id: "faster-whisper-small".to_owned(),
            state: ModelState::Ready,
            device: Some(translator_core::ComputeDevice::Cuda),
            safe_error_code: None,
        }],
        queues: ProviderQueues {
            provider_input_buffered_ms: 40,
            provider_output_buffered_ms: 0,
            queue_lag_ms: 12,
        },
        retry: None,
        safe_error: None,
    };

    let value = serde_json::to_value(health).expect("health must serialize");

    assert_eq!(value["schema_version"], "translator.provider.health.v1");
    assert_eq!(value["direction_id"], "microphone");
    assert_eq!(value["event_sequence"], 7);
    assert_eq!(value["state"], "ready");
    assert_eq!(value["queues"]["provider_input_buffered_ms"], 40);
    assert!(value.get("pcm").is_none());
    assert!(value.get("transcript").is_none());
}

#[test]
fn lifecycle_messages_use_distinct_versions_and_stable_identity() {
    let session_id = Uuid::parse_str("8ec8cb30-6881-4896-b413-7649b58cdfb2").expect("fixture UUID");
    let open = OpenProviderSession {
        schema_version: OpenSessionVersion::V1,
        session_id,
        provider_id: ProviderId::Local,
        direction_id: AudioDirection::Microphone,
        source_language: Language::Ru,
        target_language: Language::En,
        mode: TranslationMode::StreamingFirst,
        requested_input_format: pcm_format(),
        requested_output_format: pcm_format(),
        voice_profile: VoiceProfile {
            language: Language::En,
            gender: VoiceGender::Male,
            engine: VoiceEngine::Piper,
            model_path: Some("models/en_US-lessac-medium.onnx".into()),
            provider_voice_id: None,
        },
        debug_text_enabled: false,
    };
    let opened = ProviderSessionOpened {
        schema_version: SessionOpenedVersion::V1,
        session_id,
        direction_id: AudioDirection::Microphone,
        event_sequence: 1,
        negotiated_input_format: pcm_format(),
        negotiated_output_format: pcm_format(),
        capabilities: ProviderCapabilities {
            audio_output: RequiredTrue,
            transcript_delta: false,
            translation_delta: false,
            cancellation: true,
            cloud_egress: false,
        },
    };
    let close = CloseProviderSession {
        schema_version: CloseSessionVersion::V1,
        session_id,
        reason: CloseRequestReason::UserStop,
    };
    let closed = ProviderSessionClosed {
        schema_version: SessionClosedVersion::V1,
        session_id,
        direction_id: AudioDirection::Microphone,
        event_sequence: 9,
        reason: SessionCloseReason::UserStop,
    };

    assert_eq!(open.session_id, opened.session_id);
    assert_eq!(open.session_id, close.session_id);
    assert_eq!(open.provider_id, ProviderId::Local);
    assert_eq!(close.session_id, closed.session_id);
    assert_eq!(opened.capabilities.audio_output, RequiredTrue);
    assert_eq!(
        serde_json::to_value(open.schema_version).unwrap(),
        "translator.provider.open_session.v1"
    );
    assert_eq!(
        serde_json::to_value(opened.schema_version).unwrap(),
        "translator.provider.session_opened.v1"
    );
    assert_eq!(
        serde_json::to_value(close.schema_version).unwrap(),
        "translator.provider.close_session.v1"
    );
    assert_eq!(
        serde_json::to_value(closed.schema_version).unwrap(),
        "translator.provider.session_closed.v1"
    );
    assert_eq!(closed.event_sequence, 9);
    assert_eq!(opened.event_sequence, 1);
    assert!(!opened.capabilities.cloud_egress);

    let expected: serde_json::Value = serde_json::from_str(include_str!(
        "../../../tests/contract-fixtures/open_session.json"
    ))
    .expect("golden open-session payload");
    assert_eq!(serde_json::to_value(open).unwrap(), expected);

    let mut wrong_version = expected;
    wrong_version["schema_version"] = "translator.provider.health.v1".into();
    assert!(serde_json::from_value::<OpenProviderSession>(wrong_version).is_err());
    assert!(
        serde_json::from_value::<CloseProviderSession>(serde_json::json!({
            "schema_version": "translator.provider.close_session.v1",
            "session_id": session_id,
            "reason": "provider_failure"
        }))
        .is_err()
    );
}

#[test]
fn runtime_debug_update_and_probe_have_distinct_versioned_contracts() {
    let session_id = Uuid::new_v4();
    let generation_id = Uuid::new_v4();
    let update = UpdateDebugText {
        schema_version: UpdateDebugTextVersion::V1,
        session_id,
        enabled: false,
    };
    let probe = ProviderProbeRequest {
        schema_version: ProviderProbeRequestVersion::V1,
    };
    let response = ProviderProbeResponse {
        schema_version: ProviderProbeResponseVersion::V1,
        generation_id,
    };

    assert_eq!(update.session_id, session_id);
    assert!(!update.enabled);
    assert_eq!(
        serde_json::to_value(probe.schema_version).unwrap(),
        "translator.provider.probe_request.v1"
    );
    assert_eq!(
        serde_json::to_value(response.schema_version).unwrap(),
        "translator.provider.probe_response.v1"
    );
    assert_eq!(response.generation_id, generation_id);
    assert_eq!(
        serde_json::to_value(&update).unwrap(),
        serde_json::json!({
            "schema_version": "translator.provider.update_debug_text.v1",
            "session_id": session_id,
            "enabled": false
        })
    );
    assert!(
        serde_json::from_value::<UpdateDebugText>(serde_json::json!({
            "schema_version": "translator.provider.close_session.v1",
            "session_id": session_id,
            "enabled": false
        }))
        .is_err()
    );
    assert!(
        serde_json::from_value::<ProviderProbeRequest>(serde_json::json!({
            "schema_version": "translator.provider.probe_response.v1"
        }))
        .is_err()
    );
    assert!(
        serde_json::from_value::<ProviderProbeResponse>(serde_json::json!({
            "schema_version": "translator.provider.probe_request.v1",
            "generation_id": generation_id
        }))
        .is_err()
    );
}

#[test]
fn streaming_and_latency_contracts_preserve_ordering_and_policy_state() {
    let session_id = Uuid::new_v4();
    let stream_id = Uuid::new_v4();
    let utterance_id = Uuid::new_v4();
    let input = ProviderInputFrame {
        schema_version: ProviderInputVersion::V1,
        session_id,
        direction_id: AudioDirection::Microphone,
        stream_id,
        utterance_id,
        sequence: 3,
        capture_monotonic_ns: 1_000_000,
        format: pcm_format(),
        source_language: Language::Ru,
        target_language: Language::En,
        mode: TranslationMode::StreamingFirst,
        end_of_utterance: true,
        pcm: vec![0, 1, 2, 3],
    };
    let audio = ProviderAudioDelta {
        schema_version: ProviderAudioDeltaVersion::V1,
        session_id,
        direction_id: AudioDirection::Microphone,
        stream_id,
        utterance_id,
        sequence: 0,
        event_sequence: 4,
        provider_monotonic_ns: 1_100_000,
        format: pcm_format(),
        pcm: vec![4, 5, 6, 7],
    };
    let latency = ProviderLatency {
        schema_version: ProviderLatencyVersion::V1,
        session_id,
        direction_id: AudioDirection::Microphone,
        stream_id,
        event_sequence: 5,
        utterance_id: Some(utterance_id),
        asr_first_text_ms: Some(180),
        asr_final_text_ms: None,
        mt_first_text_ms: Some(75),
        tts_first_audio_ms: Some(220),
        provider_total_ms: Some(475),
    };
    let policy = LatencyPolicyState::new(
        AudioDirection::Microphone,
        TranslationMode::StreamingFirst,
        930,
        1_240,
        45,
        None,
        Some("first_audio_threshold".to_owned()),
    );

    assert_eq!(input.sequence, 3);
    assert!(input.end_of_utterance);
    assert_eq!(audio.event_sequence, 4);
    assert_eq!(latency.event_sequence, 5);
    assert_eq!(policy.current_mode, TranslationMode::StreamingFirst);

    let input_json = serde_json::to_value(input).unwrap();
    let audio_json = serde_json::to_value(audio).unwrap();
    assert_eq!(input_json["sample_rate_hz"], 16_000);
    assert_eq!(audio_json["frame_duration_ms"], 40);
    assert!(input_json.get("format").is_none());
    assert!(audio_json.get("format").is_none());

    let mut invalid_policy = serde_json::to_value(policy).unwrap();
    invalid_policy["rolling_window_seconds"] = 30.into();
    assert!(serde_json::from_value::<LatencyPolicyState>(invalid_policy).is_err());
    assert!(
        serde_json::from_value::<ProviderCapabilities>(serde_json::json!({
            "audio_output": false,
            "transcript_delta": false,
            "translation_delta": false,
            "cancellation": true,
            "cloud_egress": false
        }))
        .is_err()
    );
}

#[test]
fn pcm_format_rejects_values_outside_the_negotiated_contract() {
    assert!(PcmFormat::try_new(44_100, 1, SampleFormat::S16Le, 40).is_err());
    assert!(PcmFormat::try_new(16_000, 3, SampleFormat::S16Le, 40).is_err());
    assert!(PcmFormat::try_new(16_000, 1, SampleFormat::S16Le, 30).is_err());
    assert!(
        serde_json::from_str::<PcmFormat>(
            r#"{"sample_rate_hz":44100,"channels":1,"sample_format":"s16le","frame_duration_ms":40}"#
        )
        .is_err()
    );
}

#[test]
fn provider_error_rejects_unknown_or_content_derived_messages() {
    let marker = "private-spoken-marker";
    let unknown_field = format!(
        r#"{{"schema_version":"translator.provider.error.v1","session_id":"{}","direction_id":"speaker","event_sequence":1,"code":"provider_unavailable","retryable":true,"safe_message":"Provider is unavailable","transcript":"{}"}}"#,
        Uuid::new_v4(),
        marker
    );
    let content_derived_message = format!(
        r#"{{"schema_version":"translator.provider.error.v1","session_id":"{}","direction_id":"speaker","event_sequence":1,"code":"provider_unavailable","retryable":true,"safe_message":"{}"}}"#,
        Uuid::new_v4(),
        marker
    );
    let mismatched_static_message = format!(
        r#"{{"schema_version":"translator.provider.error.v1","session_id":"{}","direction_id":"speaker","event_sequence":1,"code":"provider_unavailable","retryable":true,"safe_message":"Required model is not loaded"}}"#,
        Uuid::new_v4()
    );

    assert!(serde_json::from_str::<PrivacySafeProviderError>(&unknown_field).is_err());
    assert!(serde_json::from_str::<PrivacySafeProviderError>(&content_derived_message).is_err());
    assert!(serde_json::from_str::<PrivacySafeProviderError>(&mismatched_static_message).is_err());
}

#[test]
fn privacy_safe_logging_projects_only_operational_error_fields() {
    let error = PrivacySafeProviderError::new(
        Uuid::new_v4(),
        AudioDirection::Speaker,
        11,
        SafeErrorCode::ProviderUnavailable,
        true,
    );

    let json = serde_json::to_string(&PrivacySafeLogEvent::from(&error)).expect("log event");

    assert!(json.contains("provider_unavailable"));
    assert!(json.contains("\"retryable\":true"));
    assert!(!json.contains("safe_message"));
    assert!(!json.contains("transcript"));
    assert!(!json.contains("pcm"));
}

#[test]
fn no_speech_has_stable_privacy_safe_contract() {
    assert_eq!(SafeErrorCode::NoSpeech.message(), "No speech was detected");
    assert_eq!(
        serde_json::to_string(&SafeErrorCode::NoSpeech).unwrap(),
        "\"no_speech\""
    );
}
